slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMapperTest.scala

```diff
@@ -76,6 +76,50 @@ class JdbcMapperTest extends AsyncTest[JdbcTestDB] {
     )
   }
 
+  def testHeadOption = {
+    import slick.util._
+
+    case class Data(a: Int, b: Int, c: Option[String])
+
+    class T(tag: Tag) extends Table[Data](tag, "T_headOption") {
+      def a = column[Int]("A")
+      def b = column[Int]("B")
+      def c = column[Option[String]]("C")
+      def * = (a, b, c).mapTo[Data]
+    }
+    val ts = TableQuery[T]
+
+    val headQ1 = ts.filter(_.a === 10).headOption
+    val headQ2 = ts.filter(_.b === 3).map(_.a + 1).headOption
+    val headQ3 = ts.filter(_.a === 2).map(_.b).headOption
+    val headQ4 = ts.filter(_.a === 1).map(_.c).headOption
+
+    seq(
+      ts.schema.create,
+      ts ++= Seq(Data(1, 2, None), Data(2, 3, Some("2")), Data(3, 4, Some("3")), Data(4, 5, None), Data(5, 6, Some("5"))),
+
+      ts.headOption.result.map(_ shouldBe Some(Data(1, 2, None))),
+
+//      headQ1.isEmpty.result.map(_ shouldBe true)
```

Comment:
There is a similar problem for subqueries. Calling `isEmpty` or `isDefined` on a query is possible independent of what you select but it would be useful to rewrite all selects to constant values in this case. Not all databases will optimize the selects away on their own.

If you run into this problem with an Option, it means that you did not get the pivoting operation right. `isEmpty` should only check the discriminator column and not care about the contents.